### PR TITLE
feat: OpenTelemetry hook

### PR DESCRIPTION
**Pull request by james-milligan**

### CODE REVIEW

File: hooks/open-telemetry/README.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    "traceId":"edc1a5f076c0afb7ea8bd2a56dfb3dd3",
+    "id":"37ea9ce3b8638962",
+    "name":"test-client.my-bool-flag",
+    "timestamp":1662116610661173,
+    "duration":100,
+    "localEndpoint":{
+        "serviceName":"hook-example"
+    },
+    "tags":{
+        "feature_flag.evaluated_value":"{\"foo\":\"bar\"}",
```
Comment (beeme1mr) [inline hooks/open-telemetry/README.md:25]: The value should not be included if a variant is available.

File: hooks/open-telemetry/pkg/otel.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+	key := fmt.Sprintf("%s.%s", hookContext.ClientMetadata().Name(), hookContext.FlagKey())
+	mw, ok := h.spans[key]
+	if !ok {
+		return errors.New("no span stored for provided hook context")
+	}
+	attributes := []attribute.KeyValue{
+		attribute.String(AttributeEvaluatedVariant, flagEvaluationDetails.ResolutionDetail.Variant),
+	}
+	fmt.Println(flagEvaluationDetails)
+	switch flagEvaluationDetails.FlagType {
```
Comment (beeme1mr) [inline hooks/open-telemetry/pkg/otel.go:108]: Here's a link to the OTel spec proposal: 

https://github.com/open-telemetry/opentelemetry-specification/pull/2529/files

The `evaluated_value` should only be used if the `evaluated_variant` is unavailable. Also, the `evaluated_value` should be cast to a string.

File: hooks/open-telemetry/pkg/otel.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+	}
+}
+
+// Finally this method is unused for this hook, spans are closed via context
+func (h Hook) Finally(hookContext of.HookContext, hookHints of.HookHints) {
+	key := fmt.Sprintf("%s.%s", hookContext.ClientMetadata().Name(), hookContext.FlagKey())
+	mw, ok := h.spans[key]
+	if ok {
+		mw.ss.span.End()
+		mw.ss.cancel()
```
Comment (beeme1mr) [inline hooks/open-telemetry/pkg/otel.go:146]: What does cancel do?

Comment (james-milligan) [reply hooks/open-telemetry/pkg/otel.go:146]: cancel closes the context linked to this request and unblocks the `goroutine` started in the `Before` method. The same happens if the parent context is closed.

File: hooks/open-telemetry/pkg/otel.go (hunk 2 above)
Comment (james-milligan) [reply hooks/open-telemetry/pkg/otel.go:111]: I have made these changes and added a unit test to ensure all returned values are of type string

**Review by beeme1mr — APPROVED** [review]
Looks good but there are a few linting issues.

File: hooks/open-telemetry/README.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+The `open telemetry hook` taps into the full hook lifecycle to write `traces` to the global `TracerProvider`.
+To ensure thread safety a lock is generated for a given `span` key, constructed from the `OpenFeature client` name and `flagKey`, any 
+spans attempting to reuse a currently active key will be blocked until the lock becomes available.
+
+By default, the hook uses an internal `context.Background()`, however a context can be provided to the hook using the 
+`hook.WithContext(context.Context)` method.
+To wait for all threads to finish processing, the `hook.Wait()` method will be used. This is managed by an internal `sync.WaitGroup{}`.
+## Example
+The following example demonstrates the use of the `OpenTelemetry hook` with the `OpenFeature golang-sdk`. The traces are sent to a `zipkin` server running at `:9411` which will receive the following trace:
```
Comment (skyerus) [inline hooks/open-telemetry/README.md:13]:
```suggestion
The following example demonstrates the use of the `OpenTelemetry hook` with the `OpenFeature go-sdk`. The traces are sent to a `zipkin` server running at `:9411` which will receive the following trace:
```

File: hooks/open-telemetry/README.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt)
+	defer cancel()
+
+	shutdown, err := initTracer(*url)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer func() {
+		if err := shutdown(ctx); err != nil {
+			log.Fatal("failed to shutdown TracerProvider: %w", err)
```
Comment (skyerus) [inline hooks/open-telemetry/README.md:93]:
```suggestion
			log.Fatalf("failed to shutdown TracerProvider: %w", err)
```

File: hooks/open-telemetry/pkg/otel.go [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+	if flagEvaluationDetails.ResolutionDetail.Variant != "" {
+		mw.ss.span.SetAttributes(
+			attribute.String(AttributeEvaluatedVariant, flagEvaluationDetails.ResolutionDetail.Variant),
+		)
+		return nil
+	}
+	var value string
+	switch flagEvaluationDetails.FlagType {
+	case of.Boolean:
+		value = fmt.Sprintf("%T", flagEvaluationDetails.Value.(bool))
```
Comment (skyerus) [inline hooks/open-telemetry/pkg/otel.go:113]: I thought `%T` printed the type?

**Review by skyerus — APPROVED** [review]

**Pull request merged**